server/api/impl/admin/synchronize.go

```diff
@@ -0,0 +1,225 @@
+			}
+			updatedAppVersionsMap[registeredApp.Manifest.AppID] = oldVersion
+		}
+	}
+
+	// Add new apps as registered
+	for appID, manifest := range adm.store.ListManifests() {
+		if _, ok := registeredAppsMap[appID]; ok {
+			continue
+		}
+		if err := adm.AddApp(manifest); err != nil {
+			return errors.Wrapf(err, "can't add new app as registered appID - %s", manifest.AppID)
+		}
+	}
+
+	registeredAppsUpgraded := adm.store.ListApps()
+
+	// call onInstanceStartup. App migration happens here
+	for _, registeredApp := range registeredAppsUpgraded {
+		if registeredApp.Status == apps.AppStatusEnabled {
+			values := map[string]string{}
+			if _, ok := updatedAppVersionsMap[registeredApp.Manifest.AppID]; ok {
+				values[oldVersionKey] = updatedAppVersionsMap[registeredApp.Manifest.AppID]
+			}
+			// Call onStartup the function of the app. It should be called only once
+			f := func() error {
+				if err := adm.call(registeredApp, registeredApp.Manifest.OnStartup, values); err != nil {
+					adm.mm.Log.Error("Can't call onStartup func of the app", "app_id", registeredApp.Manifest.AppID, "err", err.Error())
+				}
+				return nil
+			}
+			if err := adm.callOnce(f); err != nil {
+				adm.mm.Log.Error("Can't callOnce the onStartup func of the app", "app_id", registeredApp.Manifest.AppID, "err", err.Error())
+			}
+		}
+	}
+
+	return nil
+}
+
+func (adm *Admin) DeleteApp(app *apps.App) error {
+	// Call delete the function of the app
+	if err := adm.call(app, app.Manifest.OnDelete, nil); err != nil {
+		return errors.Wrapf(err, "delete failed. appID - %s", app.Manifest.AppID)
+	}
+
+	// delete oauth app
+	conf := adm.conf.GetConfig()
+	client := model.NewAPIv4Client(conf.MattermostSiteURL)
+	client.SetToken(string(adm.adminToken))
+
+	if app.OAuth2ClientID != "" {
+		success, response := client.DeleteOAuthApp(app.OAuth2ClientID)
+		if !success || response.StatusCode != http.StatusNoContent {
+			return errors.Wrapf(response.Error, "failed to delete OAuth2 App - %s", app.Manifest.AppID)
+		}
+	}
+
+	// delete app from proxy plugin, not removing the data
+	if err := adm.store.DeleteApp(app); err != nil {
+		return errors.Wrapf(err, "can't delete app - %s", app.Manifest.AppID)
+	}
+
+	adm.mm.Log.Info("Deleted the app", "app_id", app.Manifest.AppID)
+
+	return nil
+}
+
+func (adm *Admin) AddApp(manifest *apps.Manifest) error {
+	newApp := &apps.App{}
+	newApp.Manifest = manifest
+	newApp.Status = apps.AppStatusRegistered
+	if err := adm.store.StoreApp(newApp); err != nil {
+		return errors.Wrapf(err, "can't store app - %s", manifest.AppID)
+	}
+	adm.mm.Log.Info("App is registered", "app_id", manifest.AppID)
+	return nil
+}
+
+func (adm *Admin) callOnce(f func() error) error {
+	// Delete previous job
+	key := "PP_CallOnce_key"
+	if err := adm.mm.KV.Delete(key); err != nil {
+		return errors.Wrap(err, "can't delete key")
+	}
+	// Ensure all instances run this
+	time.Sleep(10 * time.Second)
```

Comment:
`Sleep` is evil. How is 10s determined? Why is it not 1s or 100s?

Reply:
Yes it is but I could not think of any other way implementing the distributed `callOnce`. If you have better suggestion I'm all ears.
I assumed 10 seconds is the **enough** time for all installations to finish the `KV.Delete` call.

Reply:
I agree that we do not have a good mechanism for a `callOnce`. One of the reasons I am suggesting we do not even address these use-cases yet. 

> I assumed 10 seconds is the enough time for all installations to finish the KV.Delete call

Right, but then if there is no downside to making the timeout longer, why is it not even longer? And if there is a downside, then 10s seems like a very long time.

Reply:
You proposed to implement it, remember? This way there will be only one `OnStartup` call for the cluster.

This is used only once on startup of the plugin. 10 second delay for running app's `OnStartup` function might not be a big problem or is it?